### PR TITLE
DEPR: deprecate Index.__getitem__ with float key

### LINKED ISSUE #34191: DEPR: deprecate Index.__getitem__ with float key
I would propose to deprecate interpreting float numbers as integers in Index `__getitem__`.

Numpy does (no longer) allow this:

```
In [1]: a = np.array([1, 2, 3])   

In [3]: a[0.0]      
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-3-2ec1d2c6d8d4> in <module>
----> 1 a[0.0]

IndexError: only integers, slices (`:`), ellipsis (`...`), numpy.newaxis (`None`) and integer or boolean arrays are valid indices
```

With a Series with `iloc`, we also don't allow this:

```
In [7]: pd.Series(a).iloc[0.0] 
...
TypeError: Cannot index by location index with a non-integer key
```

But for Index we do check if the float is integer-like and in that case convert to integer:

```
In [5]: pd.Index(a)[0.0]  
Out[5]: 1
```

However, this is not for all Index subclasses the case, eg:

```
In [9]: pd.DatetimeIndex(a)[0.0]                                                                                                                                                                                   
...
IndexError: only integers, slices (`:`), ellipsis (`...`), numpy.newaxis (`None`) and integer or boolean arrays are valid indices
```

So I think we can simply deprecate this to align Index with numpy/Series/some Index subclasses.

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
See #34191 for motivation

Closes #34191

### CODE REVIEW

**Comment by jreback** [comment]
looks good, does this close #34191?

can you add a linkin the deprecations removal issue (2.0)

File: pandas/plotting/_matplotlib/core.py [hunk 1]
```diff
@@ -1188,6 +1189,8 @@ def _post_plot_logic(self, ax, data):
         from matplotlib.ticker import FixedLocator
 
         def get_label(i):
+            if is_float(i) and i.is_integer():
+                i = int(i)
```
Comment (jorisvandenbossche) [inline pandas/plotting/_matplotlib/core.py:1193]: I am not really sure this is correct, but it's what is currently actually is being done (the conversion to int is just happening inside the `Index.__getitem__`). So this change is needed to preserve the current behaviour.

It's a bit strange that those xticks are floats, though. But maybe that's just a matplotlib behaviour.

Comment (jreback) [reply pandas/plotting/_matplotlib/core.py:1193]: this is really strange ,what is i here?

Comment (jorisvandenbossche) [reply pandas/plotting/_matplotlib/core.py:1193]: It are the values from `ax.get_xticks()`, see 5 lines below. I suppose matplotlib just always returns floats

Comment (jorisvandenbossche) [reply pandas/plotting/_matplotlib/core.py:1193]: Yes, just checked, matplotlib simply always returns floats. And it's only for the "integer floats" that we have a value in the index to put as xtick label.

File: doc/source/whatsnew/v1.1.0.rst [hunk 2]
```diff
@@ -586,6 +586,9 @@ Deprecations
 - :func:`pandas.api.types.is_categorical` is deprecated and will be removed in a future version; use `:func:pandas.api.types.is_categorical_dtype` instead (:issue:`33385`)
 - :meth:`Index.get_value` is deprecated and will be removed in a future version (:issue:`19728`)
 - :meth:`DateOffset.__call__` is deprecated and will be removed in a future version, use ``offset + other`` instead (:issue:`34171`)
+- Indexing an :class:`Index` object with a float key is deprecated, and will
+  raise an IndexError in the future. You can manually convert to an integer key
+  instead (:issue:`34191`).
```
Comment (jbrockmendel) [inline doc/source/whatsnew/v1.1.0.rst:591]: usually we dont have a trailing period.  double ticks on IndexError?

File: pandas/core/common.py [hunk 3]
```diff
@@ -144,20 +145,29 @@ def is_bool_indexer(key: Any) -> bool:
...
-def cast_scalar_indexer(val):
+def cast_scalar_indexer(val, warn_float=False):
     """
     To avoid numpy DeprecationWarnings, cast float to integer where valid.
 
     Parameters
     ----------
     val : scalar
+    warn_float : bool, default False
+        If True, raise deprecation warning for a float indexer.
```
Comment (jbrockmendel) [inline pandas/core/common.py:156]: "raise" -> "issue"

File: pandas/tests/indexes/test_indexing.py [hunk 4]
```diff
@@ -82,3 +83,16 @@ def test_contains_with_float_index(self):
         assert 1.1 in float_index
         assert 1.0 not in float_index
         assert 1 not in float_index
+
+
+@pytest.mark.parametrize(
+    "idx", [Index([1, 2, 3]), Index([0.1, 0.2, 0.3]), Index(["a", "b", "c"])]
+)
```
Comment (jbrockmendel) [inline pandas/tests/indexes/test_indexing.py:90]: could use indices fixture?

**Comment by jbrockmendel** [comment]
few comments, generally looks good

File: pandas/tests/indexes/test_indexing.py (hunk 4 above)
Comment (jorisvandenbossche) [reply pandas/tests/indexes/test_indexing.py:90]: I don't think so, as the deprecation is only for those 3 index types

**Comment by jorisvandenbossche** [comment]
> can you add a linkin the deprecations removal issue (2.0)

Added

**Pull request merged**